src/usr/local/www/interfaces_gre_edit.php

```diff
@@ -55,38 +59,80 @@
+	}
 
 	do_input_validation($_POST, $reqdfields, $reqdfieldsn, $input_errors);
 
-	if ((!is_ipaddr($_POST['tunnel-local-addr'])) ||
-	    (!is_ipaddr($_POST['tunnel-remote-addr'])) ||
-	    (!is_ipaddr($_POST['remote-addr']))) {
-		$input_errors[] = gettext("The tunnel local and tunnel remote fields must have valid IP addresses.");
+	if ($tunnel_type === '') {
+		$input_errors[] = gettext("The tunnel needs either a valid IPv4 or IPv6 tunnel configuration.");
+	}
+
+	if (((!is_ipaddr($_POST['tunnel-local-addr'])) ||
+	     (!is_ipaddr($_POST['tunnel-remote-addr']))) &&
+	    in_array($tunnel_type, array('v4', 'v4v6'))) {
+		$input_errors[] = gettext("The tunnel local and tunnel remote fields must have valid IPv4 addresses.");
+	}
+
+	if (((!is_ipaddrv6($_POST['tunnel-local-addr6'])) ||
+	     (!is_ipaddrv6($_POST['tunnel-remote-addr6']))) &&
+	     in_array($tunnel_type, array('v6', 'v4v6'))) {
+		$input_errors[] = gettext("The tunnel local and tunnel remote fields must have valid IPv6 addresses.");
+	}
+
+	if (!is_ipaddr($_POST['remote-addr'])) {
+		$input_errors[] = gettext("The remote address must be a valid IP address.");
 	}
 
 	if (!is_numericint($_POST['tunnel-remote-net'])) {
 		$input_errors[] = gettext("The GRE tunnel subnet must be an integer.");
 	}
 
-	if (is_ipaddrv4($_POST['tunnel-local-addr'])) {
+	if (!is_numericint($_POST['tunnel-remote-net6'])) {
+		$input_errors[] = gettext("The v6 GRE tunnel subnet must be an integer.");
+	}
+
+	if (in_array($tunnel_type, array('v4', 'v4v6'))) {
+		if (!is_ipaddrv4($_POST['tunnel-local-addr'])) {
+			$input_errors[] = gettext("The GRE Tunnel local address must be IPv4 if the tunnel type is IPv4.");
+		}
 		if (!is_ipaddrv4($_POST['tunnel-remote-addr'])) {
-			$input_errors[] = gettext("The GRE Tunnel remote address must be IPv4 where tunnel local address is IPv4.");
+			$input_errors[] = gettext("The GRE Tunnel remote address must be IPv4 if the tunnel type is IPv4.");
 		}
 		if ($_POST['tunnel-remote-net'] > 32 || $_POST['tunnel-remote-net'] < 1) {
-			$input_errors[] = gettext("The GRE tunnel subnet must be an integer between 1 and 32.");
+			$input_errors[] = gettext("The GRE tunnel subnet for IPv4 must be an integer between 1 and 32.");
 		}
 	}
 
-	if (is_ipaddrv6($_POST['tunnel-local-addr'])) {
-		if (!is_ipaddrv6($_POST['tunnel-remote-addr'])) {
-			$input_errors[] = gettext("The GRE Tunnel remote address must be IPv6 where tunnel local address is IPv6.");
+	if (in_array($tunnel_type, array('v6', 'v4v6'))) {
+		if (!is_ipaddrv6($_POST['tunnel-local-addr6'])) {
+			$input_errors[] = gettext("The GRE Tunnel local address must be IPv6 if the tunnel type is IPv6.");
+		}
+		if (!is_ipaddrv6($_POST['tunnel-remote-addr6'])) {
+			$input_errors[] = gettext("The GRE Tunnel remote address must be IPv6 if the tunnel type is IPv6.");
```

Comment:
See previous note, no need to mention tunnel type, just the specific field names.